# Учебная финальная работа по SQL

## Описание работы

Компания купила крупный сервис для чтения книг по подписке. Задача - проанализировать базу данных.

В базе данных информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


### Цели исследования

- Изучение базы данных;
- Анализ базы данных;
- выполнение запросов для решения каждого задания:

    - Посчитайте, сколько книг вышло после 1 января 2000 года;
    - Для каждой книги посчитайте количество обзоров и среднюю оценку;
    - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
    - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
    - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## Изучение баз данных

### Загрузка библиотек и параметров подключения

In [20]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


### Просмотр всех баз данных

Для более глубокого осмысления данных осматриваю данные в два этапа. Сначала весь файл, затем только первую строку. На первом этапе понимаю сколько строк содержит база. 

In [21]:
query = """ SELECT * FROM books """
# чтобы выполнить SQL-запрос, используем Pandas
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [22]:
query = """ SELECT * 
            FROM books
            LIMIT 1 """
# чтобы выполнить SQL-запрос, используем Pandas
pd.io.sql.read_sql(query, con = engine)

book_id  author_id         title  num_pages publication_date  publisher_id
0        1        546  'Salem's Lot        594       2005-11-01            93

In [23]:
query = """ SELECT * 
            FROM authors 
            LIMIT 1"""
# чтобы выполнить SQL-запрос, используем Pandas 636строк
pd.io.sql.read_sql(query, con = engine)

author_id      author
0          1  A.S. Byatt

In [24]:
query = """ SELECT * 
            FROM publishers 
            LIMIT 1 """
# чтобы выполнить SQL-запрос, используем Pandas 340строк
pd.io.sql.read_sql(query, con = engine)

publisher_id publisher
0             1       Ace

In [25]:
query = """ SELECT *
            FROM ratings 
            LIMIT 1"""
# чтобы выполнить SQL-запрос, используем Pandas 2793строк
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id    username  rating
0          1        1  ryanfranco       4

In [26]:
query = """ SELECT *
            FROM reviews AS r
                        LIMIT 1"""
# чтобы выполнить SQL-запрос, используем Pandas 6456строк
pd.io.sql.read_sql(query, con = engine)

review_id  book_id      username  \
0          1        1  brandtandrea   

                                                text  
0  Mention society tell send professor analysis. ...

### Выводы

В первой части работы были выполнены выгрузки баз  данных. Базы выгрузились нормально, сведения подтверждают ранее объявленные структуры. Число строк в базах:

- books - 1000 строк;
- authors - 636 строк;
- publishers - 340 строк;
- ratings - 2793 строки;
- reviews - 6456 строк.

## SQL-запросы для решения задания

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [27]:
query = """ SELECT COUNT (book_id)  
            FROM books
            WHERE publication_date > '2000-01-01' """
# чтобы выполнить SQL-запрос, используем Pandas
pd.io.sql.read_sql(query, con = engine)

count
0    819

В период с 1 января 2000 года по данным базы book_id вышло 819 книг

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [31]:
query = """ SELECT b.title,
                   COUNT (DISTINCT r1.review_id), 
                   AVG(r.rating)
            FROM reviews AS r1
            RIGHT JOIN books AS b ON r1.book_id = b.book_id
            LEFT JOIN ratings AS r ON b.book_id  = r.book_id
            GROUP BY b.book_id
            """
# чтобы выполнить SQL-запрос, используем Pandas 
pd.io.sql.read_sql(query, con = engine)#['count'].sum()

title  count       avg
0                                         'Salem's Lot      2  3.666667
1                   1 000 Places to See Before You Die      1  2.500000
2    13 Little Blue Envelopes (Little Blue Envelope...      3  4.666667
3    1491: New Revelations of the Americas Before C...      2  4.500000
4                                                 1776      4  4.000000
..                                                 ...    ...       ...
995           Wyrd Sisters (Discworld  #6; Witches #2)      3  3.666667
996                        Xenocide (Ender's Saga  #3)      3  3.400000
997                                    Year of Wonders      4  3.200000
998                        You Suck (A Love Story  #2)      2  4.500000
999  Zen and the Art of Motorcycle Maintenance: An ...      4  3.833333

[1000 rows x 3 columns]

За весь период ведения базы рейтинг был назначен на каждую книгу. Количество ревью в столбце - count, среднее значение рейтингов - в аvg.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [32]:
query = """ SELECT *
            FROM (SELECT COUNT (book_id) , p.publisher
                  FROM books AS b
                  LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id 
                  WHERE num_pages > 50 
                  GROUP BY p.publisher) AS t
            ORDER BY count DESC
            LIMIT 1     
            """
# чтобы выполнить SQL-запрос, используем Pandas
pd.io.sql.read_sql(query, con = engine)

count      publisher
0     42  Penguin Books

Издательство Penguin Books выпустило больше всего книг (42 шт.) толщиной более 50 страниц.


### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [33]:
query = """ SELECT author 
            FROM authors 
            WHERE author_id IN (SELECT author_id 
                          FROM books
                          WHERE book_id IN (SELECT book_id
                              FROM (SELECT book_id, AVG (rating)
                                    FROM ratings 
                                    WHERE book_id IN (SELECT book_id
                                                      FROM (SELECT COUNT (rating), book_id
                                                            FROM ratings 
                                                            GROUP BY  book_id
                                                            HAVING COUNT(rating) >= 50) AS t)
                                   GROUP BY book_id
                                   HAVING AVG (rating) > 0
                                   ORDER BY avg DESC
                                   LIMIT 1
                                   ) AS t
                              )
                             )
            """
# чтобы выполнить SQL-запрос, используем Pandas 
pd.io.sql.read_sql(query, con = engine)

author
0  J.K. Rowling/Mary GrandPré

Это  - J.K. Rowling/Mary GrandPré!

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [34]:
query = """ SELECT AVG (count)
            FROM (SELECT username, COUNT (review_id)
                  FROM reviews AS r
                  WHERE username IN (SELECT username
                                     FROM (SELECT COUNT (rating), username
                                     FROM ratings 
                                     GROUP BY  username
                                     HAVING COUNT(rating) > 50) AS t)
                  GROUP BY username)  AS tt               
            """
# чтобы выполнить SQL-запрос, используем Pandas 6456строк
pd.io.sql.read_sql(query, con = engine)

avg
0  24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24 и одна треть.

## Общие выводы

Все базы исправны и управляемы, данные выгружаются.

